CredentialProvider.Microsoft/CredentialProviders/VstsBuildTaskServiceEndpoint/VstsBuildTaskServiceEndpointCredentialProvider.cs

```diff
@@ -61,24 +60,92 @@ public override Task<bool> CanProvideCredentialsAsync(Uri uri)
             return Task.FromResult(true);
         }
 
-        public override Task<GetAuthenticationCredentialsResponse> HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken)
+        public override async Task<GetAuthenticationCredentialsResponse> HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
             Verbose(string.Format(Resources.IsRetry, request.IsRetry));
 
             string uriString = request.Uri.AbsoluteUri;
-            bool endpointFound = Credentials.TryGetValue(uriString, out EndpointCredentials matchingEndpoint);
-            if (endpointFound)
+            bool externalEndpointFound = ExternalCredentials.TryGetValue(uriString, out ExternalEndpointCredentials matchingExternalEndpoint);
+            if (externalEndpointFound && !string.IsNullOrWhiteSpace(matchingExternalEndpoint.Password))
             {
                 Verbose(string.Format(Resources.BuildTaskEndpointMatchingUrlFound, uriString));
                 return GetResponse(
-                    matchingEndpoint.Username,
-                    matchingEndpoint.Password,
+                    matchingExternalEndpoint.Username,
+                    matchingExternalEndpoint.Password,
                     null,
                     MessageResponseCode.Success);
             }
 
+            bool endpointFound = Credentials.TryGetValue(uriString, out EndpointCredentials matchingEndpoint);
+            if (endpointFound && !string.IsNullOrWhiteSpace(matchingEndpoint.ClientId))
+            {
+                Uri authority = await AuthUtil.GetAadAuthorityUriAsync(request.Uri, cancellationToken);
+                Verbose(string.Format(Resources.UsingAuthority, authority));
+
+                var tenantId = await AuthUtil.GetTenantIdAsync(request.Uri, cancellationToken);
+                Verbose(string.Format(Resources.UsingTenant, tenantId));
+
+                var clientCertificate = GetCertificate(matchingEndpoint);
+                Logger.Verbose(clientCertificate == null
+                    ? (Resources.ClientCertificateNotFound)
+                    : string.Format(Resources.UsingCertificate, clientCertificate.Subject));
+
+                IEnumerable<ITokenProvider> tokenProviders = await TokenProvidersFactory.GetAsync(authority);
+                tokenProviders = tokenProviders.Where(x => x.Name == "MSAL Managed Identity" || x.Name == "MSAL Service Principal").ToList();
```

Comment:
Let's add more methods to TokenProvidersFactory where necessary: e.g. GetEndpointProvidersAsync

Reply:
discussed in person, a new factory for the endpoint case to just create the providers needed and avoid the filtering out

Reply:
done